keps/sig-api-machinery/2161-apiserver-default-labels/README.md

```diff
@@ -0,0 +1,452 @@
+add nesting complexity to the network policy API, wherein certain “subsets” of
+the API were not usable together.  This is not “bad”, but it is suboptimal and
+introduces redundant semantics. This is avoidable if we could assume, in all
+cases, that namespaces have a default label, thereby leveraging existing
+functionalities offered by label selectors.
+
+There may also be many other usecases where the security or permissions boundary
+might be conveniently defined in terms of a namepsace name, and that such a
+definition would be most easily referenced by a default label.
+
+### Goals
+
+- Add the ability to select namespaces by name reliably using traditional label selector methods.
+
+### Non-Goals
+
+* Publishing name labels on all resources, this is an option we can explore later for a larger subset of fields in the Kubernetes API.
+* Exposing arbitrary fields to selectors.
+
+## Proposal
+
+We propose to label all namespaces, by default, with a reserved label of
+“kubernetes.io/metadata.name” and set it to the namespace's name. This will allow
+any resource with a `namespaceSelector` field (such as NetworkPolicy, MutatingWebhookConfiguration
+etc) to select (or exclude with `notIn`) these namespaces by explicitly matching on
+the namespace's name with traditional label selector mechanisms.
+
+- If this label is missing, it is added by the apiserver as a default.
+- If this label is deleted, it is added by the apiserver as a default.
+- defaults.go mutates namespaces on read, to always have the label `kubernetes.io/metadata.name=obj.Name`.
+- api clients always see this default label, because of this mutation.
+
+
+### User Stories (Optional)
+
+#### Avoid labels misuse
+
+NetworkPolicies allow users to open up traffic to/from their namespaces with the help of `namespaceSelector` field.
+However, there is a concern that arbitrary labels may not be the most secure way to open up traffic when it comes
+to namespace resource. In case a namespace's labels are known by others, any user can add labels at any time to any
+namespace and thus send traffic to a pod belonging to this namespace. Matching a namespace by its name, on the other
+hand, is a more reliable way to whitelist namespaces as it's much easier to specifically allow only namespaces user
+control since user knows their own namespace names.
+
+#### Lack of permissions
+
+Users want to allow DNS traffic to coreDNS pods which are created in system generated namespaces like kube-system.
+Since these system-generated namespaces do not come up with labels, users may not be authorized to add labels to
+such namespaces and thereby accessing some essential services is not straight forward.
+
+#### Simplicity
+
+It is being reported that users find it difficult to work with mutation and validation webhooks as they are not
+able to select a single namespace to enforce those controllers. Similarly, it is quite common and intuitive for
```

Comment:
we recently had a conversation about this in apimachinery and concluded that trying to segment admission webhooks to a single namespace (or small numbers of namespaces) instead of excluding a small number of namespaces to avoid a bootstrapping loop was probably a bad idea because it uses a type system inconsistently across namespaces.

Reply:
dded a comment to broaden this statement, pushing shortly